src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1561,13 +1573,20 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
     }
     if (ret.notify_waiter_count > 0) { // we have threads waiting to be notified in Object.wait()
       ObjectWaiter *waiter = mon->first_waiter();
+      jint skipped = 0;
       for (int i = 0; i < nWait; i++) {
         JavaThread *w = mon->thread_of_waiter(waiter);
+        oop thread_oop = get_vthread_or_thread_oop(w);
+        bool is_virtual = java_lang_VirtualThread::is_instance(thread_oop);
         assert(w != nullptr, "sanity check");
-        // If the thread was found on the ObjectWaiter list, then
-        // it has not been notified.
-        Handle th(current_thread, get_vthread_or_thread_oop(w));
-        ret.notify_waiters[i] = (jthread)jni_reference(calling_thread, th);
+        if (java_lang_VirtualThread::is_instance(thread_oop)) {
+          skipped++;
```

Comment:
I don't think maintaining `skipped` does anything here.

Reply:
Thank you for the question. It is needed at the line 1586 below to discount the index:
```
        if (java_lang_VirtualThread::is_instance(thread_oop)) {
          skipped++;
        } else {
          // If the thread was found on the ObjectWaiter list, then
          // it has not been notified.
          Handle th(current_thread, get_vthread_or_thread_oop(w));
          ret.notify_waiters[i - **skipped**] = (jthread)jni_reference(calling_thread, th);         <= line **1586**
        }
```

Reply:
BTW: The simplification (getting rid of local `skipped`) you requested in previous comment made damaged this fragment by making it incorrect. Here we need the `nWait` to account for virtual threads as well. Otherwise, the loop is shorted.